_posts/en/newsletters/2022-12-07-newsletter.md

```diff
@@ -0,0 +1,238 @@
+
+[Bump unconfirmed ancestor transactions to target feerate][review club 26152-2]
+is a PR by Xekyo (Murch) that improves the accuracy of the wallet's fee
+calculation in the case that unconfirmed UTXOs are selected as inputs.
+Without the PR, the fee is set too low if the
+feerates of some transactions being used as inputs are lower
+than that of the transaction being constructed. The PR fixes
+this by adding enough fee to "pay for" low-fee unconfirmed source
+transactions.
+
+Adjusting the fee to account for these ancestor transactions turns out
+to be similar to choosing which transactions to include
+in a block, so this PR adds a class called `Miniminer`.
+
+This PR review spanned two weeks,
+[Nov23][review club 26152] and [Nov30][review club 26152-2].
+
+{% include functions/details-list.md
+  q0="What problem does this PR address?"
+  a0="The wallet fee estimation doesn't take into account that it must also
+      pay for all unconfirmed ancestors with a lower feerate than the target."
+  a0link="https://bitcoincore.reviews/26152#30"
+
+  q1=”What does a transaction’s \"cluster\" consist of?”
+  a1=”A transaction's cluster is the set consisting of itself and all
+      \"connected\" transactions. A sibling, i.e. a child of
+      your parent, who is neither your ancestor nor your descendant,
+      would be part of your cluster.
+  a1link="https://bitcoincore.reviews/26152#72"
+
+  q2="This PR introduces `MiniMiner` which duplicates some of the
+      actual miner's algorithms; would it have been better to
+      unify these two implementations through refactoring?"
+  a2="We only need to operate on a cluster and not the entire mempool,
+      and don't need to apply any of the checks that `BlockAssembler`
+      does. It was also suggested to do this calculation without
+      holding the mempool lock. We'd also need to change the block
+      assembler to be tracking bump fees rather than building the
+      block template; the amount of refactoring necessary was
+      equivalent to rewriting."
+  a2link="https://bitcoincore.reviews/26152#94"
+
+  q3="Why does the MiniMiner require an entire cluster? Why can’t it
+      just use the union of each transaction’s ancestor sets?"
+  a3="Some of the ancestors may already have been paid for by some of
+      their other descendants; we don't have to pay with our transaction.
+      So we need to include these other descendants in our calculations."
+  a3link="https://bitcoincore.reviews/26152#129"
+
+  q4="If transaction X has a higher _ancestor feerate_ than independent
+      transaction Y, is it possible for a miner to prioritize Y
+      over X (that is, mine Y before X)?"
+  a4="Yes. If some of Y's low-feerate ancestors have _other_ decendants that
+      are high feerate, Y doesn't need to \"pay\" for those ancestors;
+      we can, in effect, \"remove\" those ancestors from Y's ancestor set,
+      which would then increase Y's _effective_ ancestor feerate."
```

Comment:
It's not that we can, it's that we *do*. So it's not the "effective" ancestor feerate, it's just the updated ancestor feerate.
```suggestion
      Y's ancestor set is updated to exclude those transactions,
      and Y's new ancestor feerate is higher than before.
```